### PR TITLE
Add subscribe service

### PR DESCRIPTION
**Pull request by clearloop**
## Steps

+ [x] Relayer Game
+ [x] Ethereum Relay
+ [x] Ethereum Backing

### Relayer Game

```rust
decl_event! {
	pub enum Event<T, I: Instance = DefaultInstance>
	where
		AccountId = AccountId<T>,
		RelayHeaderId = RelayHeaderId<T, I>,
	{
		/// A new relay parcel affirmed. [game id, round, index, relayer]
		Affirmed(RelayHeaderId, u32, u32, AccountId),
		/// A different affirmation submitted, dispute found. [game id]
		Disputed(RelayHeaderId),
		/// An extended affirmation submitted, dispute go on. [game id]
		Extended(RelayHeaderId),
		/// A new round started. [game id, game sample points]
		NewRound(RelayHeaderId, Vec<RelayHeaderId>),
		/// A game has been settled. [game id]
		GameOver(RelayHeaderId),
		/// Pending relay header parcel approved. [game id, reason]
		PendingRelayHeaderParcelApproved(RelayHeaderId, Vec<u8>),
		/// Pending relay header parcel rejected. [game id]
		PendingRelayHeaderParcelRejected(RelayHeaderId),
	}
}
```

https://github.com/darwinia-network/darwinia-common/blob/0c29ae4b024324adb755babce955f959f5480796/frame/bridge/relayer-game/src/lib.rs#L94-L105

### Ethereum Relay

```rust
decl_event! {
	pub enum Event<T>
	where
		<T as frame_system::Trait>::AccountId,
	{
		/// The specific confirmed parcel removed. [block id]
		RemoveConfirmedParcel(EthereumBlockNumber),
		/// EthereumReceipt verification. [account, ethereum receipt, ethereum header]
		VerifyReceipt(AccountId, EthereumReceipt, EthereumHeader),
	}
}
```
https://github.com/darwinia-network/darwinia-common/blob/0c29ae4b024324adb755babce955f959f5480796/frame/bridge/ethereum/relay/src/lib.rs#L96-L106

### Ethereum Backing

```rust
decl_event! {
	pub enum Event<T>
	where
		<T as frame_system::Trait>::AccountId,
		RingBalance = RingBalance<T>,
	{
		/// Some one redeem some *RING*. [account, amount, transaction index]
		RedeemRing(AccountId, Balance, EthereumTransactionIndex),
		/// Some one redeem some *KTON*. [account, amount, transaction index]
		RedeemKton(AccountId, Balance, EthereumTransactionIndex),
		/// Some one redeem a deposit. [account, deposit id, amount, transaction index]
		RedeemDeposit(AccountId, DepositId, RingBalance, EthereumTransactionIndex),
	}
}
```

https://github.com/darwinia-network/darwinia-common/blob/0c29ae4b024324adb755babce955f959f5480796/frame/bridge/ethereum/backing/src/lib.rs#L77-L90

### CODE REVIEW

**Comment by clearloop** [comment]
Please check the events above if there are some pending changes @AurevoirXavier

**Comment by clearloop** [comment]
Hard to test due to #40

File: src/service/subscribe/backing.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        REDEEM_KTON => redeem_kton(event),
+        REDEEM_DEPOSIT => redeem_deposit(event),
+        _ => {}
+    };
+    Ok(())
+}
+
+fn redeem_ring(event: RawEvent) {
+    if let Ok(res) = RedeemRing::<DarwiniaRuntime>::decode(&mut &event.data[..]) {
+        trace!(">> {}::{}\n\t{:?}", event.module, event.variant, res);
```
Comment (clearloop) [inline src/service/subscribe/backing.rs:27]: Log format like this prints the event structs, change it to the format needable

**Comment by HackFisher** [comment]
@clearloop How to test it? I tried to add subscribe service to listeners, but didn't print any event

File: src/service/subscribe/mod.rs [hunk 2]
```diff
@@ -1 +1,72 @@
...
+        let mut sub = EventSubscription::<DarwiniaRuntime>::new(scratch, decoder);
+        if let Some(raw) = sub.next().await {
+            if let Ok(event) = raw {
+                match event.module.as_str() {
+                    ETHEREUM_RELAY => relay::handle(event)?,
+                    ETHEREUM_BACKING => backing::handle(event)?,
+                    _ => {}
+                };
+            }
+        }
```
Comment (clearloop) [inline src/service/subscribe/mod.rs:68]: Only can prints events in the modules here.

File: src/service/subscribe/mod.rs [hunk 3]
```diff
@@ -1 +1,74 @@
...
+        let mut decoder = EventsDecoder::<DarwiniaRuntime>::new(client.metadata().clone());
+
+        // Register decoders
+        decoder.with_ethereum_backing();
+        decoder.with_ethereum_relayer_game();
+        decoder.with_ethereum_relay();
+
+        // Build subscriber
+        let mut sub = EventSubscription::<DarwiniaRuntime>::new(scratch, decoder);
+        loop {
```
Comment (clearloop) [inline src/service/subscribe/mod.rs:60]: The `loop` has been forgotten here @HackFisher

**Comment by clearloop** [comment]
> Hard to test due to #40

Now I can Affirm blocks, let me test it

**Review by HackFisher — APPROVED** [review]

**Pull request merged**